tests/test_os.sh

```diff
@@ -0,0 +1,169 @@
+# Error on unset variables
+set -u
+
+. ../liquidprompt --no-activate
+
+function test_os {
+    LP_ENABLE_COLOR=0 # Cannot test hash colors.
+    LP_ENABLE_OS=1
+
+    LP_ENABLE_OS_ARCH=1
+    LP_ENABLE_OS_FAMILY=1
+    LP_ENABLE_OS_KERNEL=1
+    LP_ENABLE_OS_DISTRIB=1
+    LP_ENABLE_OS_VERSION=1
+
+    uname() { printf 'OpenBSD i386'; }
+    _lp_os
+    assertEquals "BSD" "$lp_os_family"
+    assertEquals "OpenBSD" "$lp_os_kernel"
+    assertEquals "i386" "$lp_os_arch"
+    assertNull "$lp_os_distrib"
+    assertNull "$lp_os_version"
+    unset -f uname
+
+    uname() { printf 'MSYS ARM'; }
+    _lp_os
+    assertEquals "Windows" "$lp_os_family"
+    assertEquals "MSYS" "$lp_os_kernel"
+    assertEquals "ARM" "$lp_os_arch"
+    assertNull "$lp_os_distrib"
+    assertNull "$lp_os_version"
+    unset -f uname
+
+    uname() { printf 'Linux x86_64'; }
+    _lp_grep_fields() { lp_grep_fields=("ubuntu" "focal"); }
+    _lp_os
+    assertEquals "GNU" "$lp_os_family"
+    assertEquals "Linux" "$lp_os_kernel"
+    assertEquals "x86_64" "$lp_os_arch"
+    assertEquals "ubuntu" "$lp_os_distrib"
+    assertEquals "focal" "$lp_os_version"
+    unset -f uname
+    unset -f _lp_grep_fields
+
+    uname() { printf 'OSCO 486'; }
+    _lp_grep_fields() { lp_grep_fields=("urs"); }
+    _lp_os
+    assertNull "$lp_os_family"
+    assertNull "$lp_os_kernel"
+    assertEquals "486" "$lp_os_arch"
+    assertNull "$lp_os_distrib"
+    assertNull "$lp_os_version"
+    unset -f uname
+    unset -f _lp_grep_fields
+
+    uname() { printf 'Linux x86_64'; }
+    _lp_grep_fields() { lp_grep_fields=("ubuntu" "focal"); }
+    LP_MARK_OS=("FreeBSD" "NOPE" "x86" "NOPE" "GNU" "G" "Linux" "L" "x86_64" "6" "ubuntu" "U" "focal" "F")
+
+    LP_MARK_OS_SEP="+"
+    LP_ENABLE_OS_ARCH=1
+    LP_ENABLE_OS_FAMILY=1
+    LP_ENABLE_OS_KERNEL=1
+    LP_ENABLE_OS_DISTRIB=1
+    LP_ENABLE_OS_VERSION=1
+    _lp_os_color
+    assertEquals "6+G+L+U+F" "$lp_os_color"
```

Comment:
For some unknown reason this test errors with a stack recusion overflow on Zsh 5.0:
```
+test_os:60> _lp_os_color
+_lp_os_color:5> lp_os_color=''
+_lp_os_color:6> local lp_os_arch lp_os_family lp_os_kernel lp_os_distrib lp_os_version
+_lp_os_color:8> _lp_os
+_lp_os:2> ((  LP_ENABLE_OS  ))
+_lp_os:5> lp_os_arch=''
+_lp_os:6> lp_os_family=''
+_lp_os:7> lp_os_kernel=''
+_lp_os:8> lp_os_distrib=''
+_lp_os:9> lp_os_version=''
+_lp_os:11> local uname_sm kernel
+_lp_os:14> ((  LP_ENABLE_OS_ARCH  ))
+_lp_os:16> uname_sm=+_lp_os:16> uname -s -m
+uname:0> printf 'Linux x86_64'
+_lp_os:16> uname_sm='Linux x86_64'
+_lp_os:17> IFS=' ' +_lp_os:17> read -r kernel lp_os_arch
+_lp_os:22> ((  LP_ENABLE_OS_KERNEL
       || LP_ENABLE_OS_FAMILY
       || LP_ENABLE_OS_DISTRIB
       || LP_ENABLE_OS_VERSION  ))
+_lp_os:27> case Linux (FreeBSD)
+_lp_os:27> case Linux (DragonFly)
+_lp_os:27> case Linux (OpenBSD)
+_lp_os:27> case Linux (Darwin)
+_lp_os:27> case Linux (SunOS)
+_lp_os:27> case Linux (CYGWIN*)
+_lp_os:27> case Linux (MSYS*)
+_lp_os:27> case Linux (MINGW*)
+_lp_os:27> case Linux (Linux)
+_lp_os:61> ((  LP_ENABLE_OS_KERNEL  ))
+_lp_os:61> lp_os_kernel=Linux
+_lp_os:62> ((  LP_ENABLE_OS_FAMILY  ))
+_lp_os:63> [[ Linux == Linux ]]
+_lp_os:64> osfamily=+_lp_os:64> uname -o
+uname:0> printf 'Linux x86_64'
+_lp_os:64> osfamily='Linux x86_64'
+_lp_os:65> [[ 'Linux x86_64' == Android ]]
+_lp_os:70> lp_os_family=GNU
+_lp_os:73> ((  LP_ENABLE_OS_DISTRIB && LP_ENABLE_OS_VERSION  ))
+_lp_os:74> _lp_grep_fields /etc/os-release '=' ID VERSION_CODENAME
+_lp_grep_fields:0> lp_grep_fields=( ubuntu focal )
+_lp_os:75> lp_os_distrib=ubuntu
+_lp_os:76> lp_os_version=focal
+_lp_os:91> [[ -z x86_64GNULinuxubuntufocal ]]
+_lp_os_color:10> local lp_substitute lp_hash_color lp_join
+_lp_os_color:21> [[ -n GNU ]]
+_lp_os_color:22> _lp_substitute GNU FreeBSD NOPE x86 NOPE GNU G Linux L x86_64 6 ubuntu U focal F
+_lp_substitute:3> lp_substitute=''
+_lp_substitute:4> local 'string=GNU'
+local:1> [[ -n x86_64 ]]
+local:2> _lp_substitute x86_64 FreeBSD NOPE x86 NOPE GNU G Linux L x86_64 6 ubuntu U focal F
+_lp_substitute:3> lp_substitute=''
+_lp_substitute:4> local 'string=x86_64'
+local:1> [[ -n x86_64 ]]
+local:2> _lp_substitute x86_64 FreeBSD NOPE x86 NOPE GNU G Linux L x86_64 6 ubuntu U focal F
+_lp_substitute:3> lp_substitute=''
+_lp_substitute:4> local 'string=x86_64'
...
local:2: job table full or recursion limit exceeded
```

Reply:
That other `active=()` you spotted (which I missed) was apparently the cause of this crash, because now it works.